esp-hal/src/soc/esp32/psram.rs

```diff
@@ -9,53 +9,126 @@
 //! storage capacity and improved performance in certain applications.
-//!
-//! The `PSRAM` module is accessed through a virtual address, defined as
-//! `PSRAM_VADDR`. The starting virtual address for the PSRAM module is
-//! 0x3F800000. The `PSRAM` module size depends on the configuration specified
-//! during the compilation process. The available `PSRAM` sizes are `2MB`,
-//! `4MB`, and `8MB`.
-//!
-//! NOTE: If you want to use `PSRAM` on `ESP32` or `ESP32-S3`, it'll work only
-//! in `release` mode.
 
-/// The starting virtual address of the PSRAM (Pseudo-SRAM) region in memory.
-pub const PSRAM_VADDR_START: usize = 0x3F800000;
+const EXTMEM_ORIGIN: usize = 0x3F800000;
+
+/// Cache Speed
+#[derive(PartialEq, Eq, Clone, Copy, Debug, Default)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[allow(missing_docs)]
+pub enum PsramCacheSpeed {
+    #[default]
+    PsramCacheF80mS40m = 0,
+    PsramCacheF40mS40m,
+    PsramCacheF80mS80m,
+}
 
-/// Retrieves the starting virtual address of the PSRAM (Pseudo-SRAM) region in
-/// memory.
-pub fn psram_vaddr_start() -> usize {
-    PSRAM_VADDR_START
+/// Size of PSRAM
+///
+/// [PsramSize::AutoDetect] will try to detect the size of PSRAM
+#[derive(Copy, Clone, Debug, Default, PartialEq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum PsramSize {
+    /// Detect PSRAM size
+    #[default]
+    AutoDetect,
+    /// A fixed PSRAM size
+    Size(usize),
 }
 
-cfg_if::cfg_if! {
-    if #[cfg(feature = "psram-2m")] {
-        const PSRAM_SIZE: u32 = 2;
-    } else if #[cfg(feature = "psram-4m")] {
-        const PSRAM_SIZE: u32 = 4;
-    } else if #[cfg(feature = "psram-8m")] {
-        const PSRAM_SIZE: u32 = 8;
-    } else {
-        const PSRAM_SIZE: u32 = 0;
+impl PsramSize {
+    fn get(&self) -> usize {
+        match self {
+            PsramSize::AutoDetect => 0,
+            PsramSize::Size(size) => *size,
+        }
+    }
+
+    fn is_auto(&self) -> bool {
+        matches!(self, PsramSize::AutoDetect)
     }
 }
 
-/// The total size of the PSRAM (Pseudo-SRAM) in bytes.
-pub const PSRAM_BYTES: usize = PSRAM_SIZE as usize * 1024 * 1024;
+/// PSRAM configuration
+#[derive(Copy, Clone, Debug, Default)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub struct PsramConfig {
+    /// PSRAM size
+    pub size: PsramSize,
+    /// Cache speed
+    pub cache_speed: PsramCacheSpeed,
+}
 
 /// Initializes the PSRAM memory on supported devices.
-#[cfg(any(feature = "psram-2m", feature = "psram-4m", feature = "psram-8m"))]
-pub fn init_psram(_peripheral: impl crate::peripheral::Peripheral<P = crate::peripherals::PSRAM>) {
-    utils::psram_init();
-    utils::s_mapping(PSRAM_VADDR_START as u32, PSRAM_BYTES as u32);
+///
+/// Returns the start of the mapped memory and the size
+pub fn init_psram(
+    _peripheral: impl crate::peripheral::Peripheral<P = crate::peripherals::PSRAM>,
+    config: PsramConfig,
+) -> (*mut u8, usize) {
```

Comment:
Isn't this return value just a slice? i.e.
```suggestion
) -> &'static mut [u8] {
```

Reply:
Risky, if we ever end up statically allocating anything in PSRAM, `&mut` might allow breaking aliasing rules.

Reply:
I agree but at that point this `init_psram` would have to go away completely for static linking, since you have to initialize PSRAM before main.

At least I'd expect some `split_at_mut` to have happened before this slice is returned.

Reply:
My thinking is that to implement s.th. like static linking, dealing with raw-addresses will be much nicer than dealing with a slice (since for copying `.data` and zeroing `.bss` will deal a lot with LMA/VMA - in case of `.text` it even needs to translate ibus/dbus addresses) and I assume that will build upon this (i.e. not built into init-psram) - later at some stage we might want to have a way to get a slice of the remaining heap (but MaybeUninit)

Reply:
Imo this is best left as a pointer + size for now